### PR TITLE
Add U2f Login Support

### LINKED ISSUE #53: Add U2F verification to further secure the server login
Currently, there is 2FA authentication which is a great way to secure online accounts, but I would like to suggest adding U2F authentication (Ledger/Trezor support it) to further enhance the security of the server login.

### PR DESCRIPTION
**Pull request by Kukks**
Closes #53

### CODE REVIEW

**Comment by pavlenex** [comment]
Tests are failing, but I've tested this long-awaited PR on over 10 U2F devices.

tACK.
Devices below work out of the box on both Chrome and Firefox (all latest).
U2F BTCPay

One exception is Nano S that has already known Chrome problem, so you need to quickly confirm when prompted, but that one should be fixed soon by Ledger.

This was so much fun to test.

**Comment by Kukks** [comment]
I actually did all my dev work and testing on a Nano S so it was not that bad with it :)

File: BTCPayServer/Views/Account/LoginWithU2F.cshtml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        2: "Bad request error, try again",
+        3: "This key isn't supported, please try another one",
+        4: 'The device is already registered, please login',
+        5: 'Authentication timed out. Please reload to try again.'
+    };
+    setTimeout(function() {
+        window.u2f.sign(
+            "@Model.AppId",
+            "@Model.Challenge",
+            @Html.Raw(@Model.Challenges), function (data) {
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Account/LoginWithU2F.cshtml]: JS injection possible

File: BTCPayServer/Views/Account/LoginWithU2F.cshtml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                console.log("call back data: ", data);
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Account/LoginWithU2F.cshtml]: no logs

**Comment by NicolasDorier** [comment]
U2F is so simple that I think you can remove the dependency to `U2F.Core`. IMHO, this is not worth it.

**Comment by NicolasDorier** [comment]
@pavlenex you can test it?

**Comment by Kukks** [comment]
> U2F is so simple that I think you can remove the dependency to `U2F.Core`. IMHO, this is not worth it.

I do not have experience with the lower level crypto stuff for the spec so I prefer trusting a tested package

File: BTCPayServer/Views/Account/LoginWithU2F.cshtml (hunk 1 above)
Comment (Kukks) [reply BTCPayServer/Views/Account/LoginWithU2F.cshtml]: What should I do for this then? I don't think injection here would be harmful?

**Comment by pavlenex** [comment]
My suggestions are all UX related.

## Setup Page (Registring U2F Device)
Capture
1. Add [font-awesome loading icon ](https://fontawesome.com/how-to-use/on-the-web/styling/animating-icons)next to Registring U2F Device, to indicate something is happening, loading, waiting.
2. Replace description text, with `Insert your U2F device or a hardware wallet into your computer's USB port. If it has a button, tap on it`

## Login page (U2F only enabled)
Same as for point 1. and 2. in the setup page.

## Login page (Both U2F and 2FA enabled)
Personally, if both are enabled, I'd show U2F with a link at the bottom `Try alternative authentication method`, but it really is not a big deal.

File: BTCPayServer/Controllers/AccountController.cs [hunk 3]
```diff
@@ -91,8 +97,39 @@ public async Task<IActionResult> Login(LoginViewModel model, string returnUrl =
...
+
+                        return View("SecondaryLogin", new SecondaryLoginViewModel()
+                        {
+                            LoginWith2FaViewModel = twoFModel,
+                            LoginWithU2FViewModel = await BuildU2FViewModel(model.RememberMe, user)
+                        });
+                    }
+                    else
+                    {
+                      var incrementAccessFailedResult = await  _userManager.AccessFailedAsync(user);
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/AccountController.cs]: If you plan doing this, then you need to document to user how to unlock his account by modifying the DB directly.

File: BTCPayServer/Controllers/AccountController.cs [hunk 4]
```diff
@@ -123,6 +162,63 @@ public async Task<IActionResult> Login(LoginViewModel model, string returnUrl =
             return View(model);
         }
 
+        private async Task<LoginWithU2FViewModel> BuildU2FViewModel(bool rememberMe, ApplicationUser user)
+        {
+            if (Request.IsHttps)
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/AccountController.cs]: If onion is also fine. I think there is a method somewhere I created search for "IsSecure" in the code.

File: BTCPayServer/Controllers/AccountController.cs (hunk 3 above)
Comment (Kukks) [reply BTCPayServer/Controllers/AccountController.cs]: The lockout is interval based and already enabled on normal login flow

Comment (Kukks) [reply BTCPayServer/Controllers/AccountController.cs]: I think after 10 mins the user is no longer locked using ASP Identity

**Comment by NicolasDorier** [comment]
Need tests

File: BTCPayServer/Views/Manage/AddU2FDevice.cshtml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            2: "Bad request error, try again",
+            3: "This key isn't supported, please try another one",
+            4: 'The device is already registered, please login',
+            5: 'Authentication timed out. Please reload to try again.'
+        };
+
+        setTimeout(function() {
+                var request = { "challenge": "@Model.Challenge", "version": "@Model.Version", "appId": "@Model.AppId" };
+                var registerRequests = [{version: request.version, challenge: request.challenge}];
+                console.log(request);
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Manage/AddU2FDevice.cshtml]: don't log

File: BTCPayServer/Views/Account/LoginWithU2F.cshtml (hunk 1 above)
Comment (NicolasDorier) [reply BTCPayServer/Views/Account/LoginWithU2F.cshtml]: I don't remember how I do, search `srvModel` in the code, I show how to get a json object properly deserialized.

File: BTCPayServer/Controllers/AccountController.cs [hunk 6]
```diff
@@ -123,6 +162,63 @@ public async Task<IActionResult> Login(LoginViewModel model, string returnUrl =
...
+            }
+            try
+            {
+                if (await _u2FService.AuthenticateUser(viewModel.UserId, viewModel.DeviceResponse))
+                {
+                    await _signInManager.SignInAsync(user, viewModel.RememberMe, "U2F");
+                    _logger.LogInformation("User logged in.");
+                    return RedirectToLocal(returnUrl);
+                }
+                throw new Exception("Invalid login attempt.");
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/AccountController.cs]: wtf? just don't throw anything here and do what is in the catch next.

**Pull request merged**